src/python/pants/goal/run_tracker.py

```diff
@@ -390,3 +413,40 @@ def shutdown_worker_pool(self):
     N.B. This exists only for internal use and to afford for fork()-safe operation in pantsd.
     """
     SubprocPool.shutdown(self._aborted)
+
+  def report_target_info(self, scope, target, keys, val):
+    """Add target information to run_info under target_data.
+
+    Will Recursively construct a nested dict with the keys provided.
+
+    :param string scope: The scope for which we are reporting the information.
+    :param string target: The target for which we want to store information.
+    :param list of string keys: The keys that will be recursively
+           nested and pointing to the information being stored.
+    :param dict or string val: The value of the information being stored.
+
+    :API: public
+    """
+    def create_dict_with_nested_keys_and_val(all_keys, val_item, index):
```

Comment:
I think this doesn't need the `all_keys` param at all? It can access `keys` directly from its closure.